lib/codegen/src/constant_folding.rs

```diff
@@ -0,0 +1,118 @@
+//! Fold operations on constants at compile time.
+
+use cursor::{Cursor, FuncCursor};
+use ir::{self, InstBuilder};
+// use timing;
+
+/// Fold operations on constants.
+///
+/// It's important to note that this will not remove unused constants. It's
+/// assumed that the DCE pass will take care of them.
+pub fn fold_constants(func: &mut ir::Function) {
+    // let _tt = timing::constant_folding();
+    let mut pos = FuncCursor::new(func);
+
+    while let Some(_ebb) = pos.next_ebb() {
+        while let Some(inst) = pos.next_inst() {
+            use ir::instructions::Opcode::*;
+            match pos.func.dfg[inst].opcode() {
```

Comment:
An organization that I think would be advantageous would be to match on the `InstructionFormat`, which is `pos.func.dfg[inst]`, rather than pulling out the opcode right away. Then there will be one match arm for all `ir::InstructionData::Binary` opcodes, and then we can have one `fold_numerical_binary` call and pass it the opcode. This seems appealing because it seems like it could let us keep the code for manipulating the IR separate from the folding arithmetic logic. If the ultimate folding function for binary operators could be a function which takes an opcode and two i64's and returns an i64 (or maybe an Option<i64> because trapping operators), that'd (a) make it really easy to write unit tests for the folding arithmetic, and (b) make it possible to reuse the folding arithmetic for... other fun things in the future, like IR interpreters, or auto-folding InstBuilders :-).

Reply:
Oh boy, didn't even think about an auto-folding `InstBuilder`. That would be super cool, and would probably remove most of the overhead of a constant folding pass.